### PR TITLE
Refactor to echo and add metrics

### PR DESCRIPTION
**Pull request by andy-v-h**
* switches from `gorilla/mux` to `labstack/echo` as the http router
* add Prometheus metrics
* ensures that startup context is passed to the signal catch and that request context are passed to S3 operations

Large sections are refactored to integrate into the `echo` http framework. This includes removing sections that would be handled by middleware. focusing the use to a more direct s3 -> https interface.

### CODE REVIEW

File: internal/controllers/s3.go [hunk 1]
```diff
@@ -186,7 +140,7 @@ func setStrHeader(w http.ResponseWriter, key string, value *string) {
 
 func setIntHeader(w http.ResponseWriter, key string, value *int64) {
 	if value != nil && *value > 0 {
-		w.Header().Add(key, strconv.FormatInt(*value, 10))
+		w.Header().Add(key, strconv.FormatInt(*value, 10)) // nolint
```
Comment (sfunkhouser) [inline internal/controllers/s3.go:143]: What are you not linting?

Comment (andy-v-h) [reply internal/controllers/s3.go:143]: `10`. it's magic number

File: internal/controllers/s3.go [hunk 2]
```diff
@@ -36,114 +35,69 @@ func AwsS3Get(w http.ResponseWriter, r *http.Request) {
...
-
-			return
+			return s3listFiles(e, c.S3Bucket, c.S3KeyPrefix+path)
 		}
 
 		path += c.IndexDocument
 	}
 	// Get a S3 object
-	obj, err := client.S3get(c.S3Bucket, c.S3KeyPrefix+path, rangeHeader)
+	obj, err := service.S3get(req.Context(), c.S3Bucket, c.S3KeyPrefix+path, rangeHeader)
```
Comment (sfunkhouser) [inline internal/controllers/s3.go:56]: does there need to be a 
```
defer obj.Body.Close()
```
here?

**Review by goshlanguage — APPROVED** [review]
lgtm

**Review by sfunkhouser — APPROVED** [review]

File: internal/controllers/s3.go (hunk 2 above)
Comment (goshlanguage) [reply internal/controllers/s3.go:56]: I don't think that would be needed, the func signature returns data, and not an open handler:
```
func S3get(ctx context.Context, bucket, key string, rangeHeader *string) (*s3.GetObjectOutput, error) {
	req := &s3.GetObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(key),
		Range:  rangeHeader,
	}

	return s3.New(c.Session).GetObjectWithContext(ctx, req)
}
```

The example from the docs doesn't appear to be defering either:
```
svc := s3.New(sess, &aws.Config{
   	DisableRestProtocolURICleaning: aws.Bool(true),
})
out, err := svc.GetObject(&s3.GetObjectInput {
   	Bucket: aws.String("bucketname"),
    	Key: aws.String("//foo//bar//moo"),
})
```
src: https://pkg.go.dev/github.com/aws/aws-sdk-go/service/s3#hdr-Automatic_URI_cleaning

**Pull request merged**